0004-predicate-types.md

```diff
@@ -0,0 +1,192 @@
+  and a subset type of the same name with a call to the predicate for constraint, with the witness provided.
+  We also need a predicate to be able to have the same name as a type, for example by internally prefixing types with "t_"
+
+There is nothing more to implement for the verifier and compiler, since this is desugarding.
+The last example above would be rewritten:
+
+```dafny
+predicate Odd(x: int) {
+  x % 2 == 1
+}
+type Odd = x: int | Odd(x) witness 1
+
+method Main() {
+  var i:int := f();
+  var j: Odd := if Odd(i) then i else 1;
+  j := j + 1; // You'd get an error here, cannot assign something not Odd to something that is Odd
+  j := j + 2; // Here you get no error, assuming j is odd.
+}
+```
+
+which would automatically satisfy all 
+
+## Further notes
+
+* We would still be able to use `{:opaque}` and use `reveal` because that annotation can be directly added after `predicate`.
+* It would be possible to define `predicate newtype` the same way, and it would have the same advantages.
+* In a second milestone, we could define another kind of `predicate type` inside a class or a datatype,
+  where the type is statically accessible, but predicate is not and has no argument. For example:
+  ```dafny
+  datatype Boxed = BoxedNumber(n: int) {
+    predicate type Odd() {
+      this.n % 2 == 1
+    }
+  }
+  ... var j: Boxed.Odd = BoxedNumber(1) ...
+  ```
+  would become
+  ```dafny
+  datatype Boxed = BoxedNumber(n: int) {
+    predicate Odd() {
+      this.n % 2 == 1
+    }
+    type Odd = s: Boxed | s.Odd()
+  }
+  ... var j: Boxed.Odd = BoxedNumber(1) ...
+  ```
+  Dafny does not support nested types (which would automatically be static) so that would be a huge refactoring.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It feels odd to have restrictions on a predicate so that it becomes a type.
+* It might be even more confusing for newcomers.
+* It feels weird that the type is statically accessible in a class but there is no "static" keyword in front of `predicate type`.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- For every subset type `MyType`, we could also generate by default a predicate with the name `MyType?` that has the same constraint.
+  That might actually be even simpler, but then there would be no control over the predicate like `{:opaque}` and `reveal`.
+- `type predicate` instead of `predicate type`? What should be the focus? I think the second one is more English.
+- Should the name of the predicate and the type be the same? We could also enfore the name of the predicate to end with `?`,
+  so that the type doesn't have the `?`. For example, `predicate type Even?` that becomes `predicate Even?` and `type Even`.
+
+# Prior art
+[prior-art]: #prior-art
+
+This feature to combine both the predicate and a subset type in a single construct does not exist in other programming languages, as far as I know.
```

Comment:
A related issue are Dafny's `least predicate`s, which are predicates in Dafny and are data in Coq (declared as `Inductive`).

The theorem prover PVS has predicate subtypes. I'm unsure, but it may also have a syntax that lets you combine the predicate and the type (either like the `predicate type` here or as an `Odd?` predicate that is automatically generated from a type `Odd`, like in the example below).